### PR TITLE
Allocate a bit less

### PR DESCRIPTION
**Pull request by roberth**
# Motivation

Check off a few dumb optimizations and a `!!!`, so that we allocate a little less in
- `readDir`
- `functionArgs`
- similarly, the `Bindings::alloc` call sites that are invoked more than a couple of times.

Also fixes a trivial warning emitted by clang.

# Context

Like `vNull` which snuck into #10251.

# Priorities and Process

Add :+1: to [pull requests you find important](https://github.com/NixOS/nix/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc).

The Nix maintainer team uses a [GitHub project board](https://github.com/orgs/NixOS/projects/19) to [schedule and track reviews](https://github.com/NixOS/nix/tree/master/maintainers#project-board-protocol).

### CODE REVIEW

File: src/libexpr/eval.hh [hunk 1]
```diff
@@ -187,10 +187,35 @@ public:
...
+    Value vFalse;
+
+    /** `"regular"` */
+    Value vStringRegular;
+    /** `"directory"` */
+    Value vStringDirectory;
+    /** `"symlink"` */
+    Value vStringSymlink;
+    /** `"unknown"` */
+    Value vStringUnknown;
```
Comment (Ericson2314) [inline src/libexpr/eval.hh:217]: Maybe make the doc refer to the builtin for which this is for?

**Review by Ericson2314 — APPROVED** [review]

File: src/libexpr/eval.hh (hunk 1 above)
Comment (roberth) [reply src/libexpr/eval.hh:217]: That info could bitrot, and is easy enough to find reliably. It should take about 1 second in your IDE/editor to find the usages or occurrences.

Comment (Ericson2314) [reply src/libexpr/eval.hh:217]: Guess I need to improve my editor setup :)

Comment (roberth) [reply src/libexpr/eval.hh:217]: In vscode, either:
- Shift+F12 for references (needs language server)
- Double click and Ctrl+Shift+F for dumb grep

I'm sure something similar is available or achievable elsewhere.

**Pull request merged**